core/templates/dev/head/collection_player/CollectionDataService.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to retrieve information about collections from the
+ * backend.
+ *
+ * @author henning.benmax@gmail.com (Ben Henning)
+ */
+
+// TODO(bhenning): For preview mode, this service should be omitted. Instead,
+// a separate collectionDataService should be created which serves a local copy
```

Comment:
I'm not sure that pattern applies. That pattern seems to simulating inheritance to minimize the amount of code needed for implementation. We need a substitution pattern, though. The frontend is expecting some sort of collectionDataService instance that has a specific interface. However, the collection player can exist in two possible contexts: a learner context and an editor/creator context. These two different contexts lead to different behaviors for the collectionDataService.

In my opinion, the collection player should be agnostic of which context it's in, and so should the collectionDataService. We gain a lot of flexibility if we have a substitution pattern for services, because it allows us to reuse controllers and other services fluidly in different contexts. In this case, we switch out the standard learner context collectionDataService which communicates to the backend with one which operates on a local copy of a collection, avoiding the need for a backend call altogether when the context is the collection editor.